### PR TITLE
ORV2-2293 Create/Manage Credit Accounts

### PR DESCRIPTION
**Pull request by glen-aot**
# Description

Implements the create/hold/close and manage users interface for FIN users accessing credit accounts. 

This covers the following tickets:  [ORV2-2293: FIN Create credit account](https://jira.th.gov.bc.ca/secure/RapidBoard.jspa?rapidView=10395&projectKey=ORV2&view=detail&selectedIssue=ORV2-2293&quickFilter=11949), [ORV2-2350: FIN hold credit account](https://jira.th.gov.bc.ca/secure/RapidBoard.jspa?rapidView=10395&projectKey=ORV2&view=detail&selectedIssue=ORV2-2350&quickFilter=11949),[ ORV2-2352: FIN close credit account](https://jira.th.gov.bc.ca/secure/RapidBoard.jspa?rapidView=10395&projectKey=ORV2&view=detail&selectedIssue=ORV2-2352&quickFilter=11949) and[ ORV2-2354: FIN manage credit account users](https://jira.th.gov.bc.ca/secure/RapidBoard.jspa?rapidView=10395&projectKey=ORV2&view=detail&selectedIssue=ORV2-2354&quickFilter=11949).

Fixes [ORV-2293, ORV2-2350, ORV2-2352, ORV2-2354]

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have already been accepted and merged

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://onroutebc-1472-frontend.apps.silver.devops.gov.bc.ca)
- [Vehicles](https://onroutebc-1472-vehicles.apps.silver.devops.gov.bc.ca/api)
- [Dops](https://onroutebc-1472-dops.apps.silver.devops.gov.bc.ca/api)
- [Policy](https://onroutebc-1472-policy.apps.silver.devops.gov.bc.ca/api)
- [Scheduler](https://onroutebc-1472-scheduler.apps.silver.devops.gov.bc.ca/api)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/onroutebc/actions/workflows/analysis.yml)

After merge, new images are promoted to:
- [Merge Workflow](https://github.com/bcgov/onroutebc/actions/workflows/merge.yml)

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://onroutebc-1472-frontend.apps.silver.devops.gov.bc.ca)
- [Vehicles](https://onroutebc-1472-vehicles.apps.silver.devops.gov.bc.ca/api)
- [Dops](https://onroutebc-1472-dops.apps.silver.devops.gov.bc.ca/api)
- [Policy](https://onroutebc-1472-policy.apps.silver.devops.gov.bc.ca/api)
- [Scheduler](https://onroutebc-1472-scheduler.apps.silver.devops.gov.bc.ca/api)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/onroutebc/actions/workflows/analysis.yml)

After merge, new images are promoted to:
- [Merge Workflow](https://github.com/bcgov/onroutebc/actions/workflows/merge.yml)

### CODE REVIEW

File: frontend/src/common/helpers/tableHelper.ts
Comment (zgong-gov) [inline frontend/src/common/helpers/tableHelper.ts]: Is it possible to override these values in the component you want to use, rather than removing it here, since this could be reused by other components as well (and can have unintended consequences)?

File: frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx
Comment (zgong-gov) [inline frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx:80]: Do we still need feature flags for credit accounts (now that the feature is completed)?

File: frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx
Comment (zgong-gov) [inline frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx]: Even though it's guaranteed that companyId exists, we typically don't use the "!" syntax. Use either ```Number(companyId)``` or ```getDefaultRequiredVal(0, companyId)``` instead

File: frontend/src/features/settings/apiManager/endpoints/endpoints.ts
Comment (zgong-gov) [inline frontend/src/features/settings/apiManager/endpoints/endpoints.ts]: For all of these places here that need "getCompanyIdFromSession", replace it with a companyId param instead, ie. ```GET_CREDIT_ACCOUNT_USERS: (creditAccountId: number, companyId: number | string) => ...```, and pass in the companyId from the caller. It's not a good idea to simply grab the companyId from the session anywhere throughout the app.

Comment (zgong-gov) [reply frontend/src/features/settings/apiManager/endpoints/endpoints.ts]: Call getCompanyIdFromSession inside the parent credit account page component, and pass the companyId as param to these route, as well as the api methods and hooks that uses these routes

File: frontend/src/features/settings/components/creditAccount/AccountDetails.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/AccountDetails.tsx:193]: Is this feature complete? Do we still need mock values here?

File: frontend/src/features/settings/components/creditAccount/ActivityTable.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/ActivityTable.tsx]: Use ```...getDefaultRequiredVal([], creditAccount?.creditAccountActivities)``` instead

File: frontend/src/features/settings/components/creditAccount/ActivityTable.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/ActivityTable.tsx]: Maybe it's a good idea to create an enum type for creditAccountActivityType, ie.
```
const CREDIT_ACCOUNT_ACTIVITY_TYPES = { OPENED: "OPENED", ... } as const;
type CreditAccountActivityType = typeof CREDIT_ACCOUNT_ACTIVITY_TYPES[keyof typeof CREDIT_ACCOUNT_ACTIVITY_TYPES];
```
and use it here as: ```activity.creditAccountActivityTypes !== CREDIT_ACCOUNT_ACTIVITY_TYPES.OPENED```

File: frontend/src/features/settings/components/creditAccount/ActivityTable.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/ActivityTable.tsx]: Once again here, anywhere that has "??" syntax, use getDefaultRequiredVal instead, ie. ```data: getDefaultRequiredVal([], creditAccountActivities)```

File: frontend/src/features/settings/components/creditAccount/AddUser.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/AddUser.tsx]: It might be better here to rename data as companyInfo, so that it's more clear what data is supposed to be throughout the code, ie. ```const { data: companyInfo, ... }```

File: frontend/src/features/settings/components/creditAccount/AddUser.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/AddUser.tsx]: I'm confused here, is the behaviour supposed to be different when data doesn't exist (ie. undefined, as in during the loading state) and when it's empty?

File: frontend/src/features/settings/components/creditAccount/AddUserModal.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/AddUserModal.tsx]: Maybe consider using an enum type here as well, ie.
```
const CREDIT_ACCOUNT_USER_TYPE = { HOLDER: "HOLDER", ... } as const;
type CreditAccountUserType = typeof CREDIT_ACCOUNT_USER_TYPE[keyof typeof CREDIT_ACCOUNT_USER_TYPE];
```
and use it here as ```user.userType === CREDIT_ACCOUNT_USER_TYPE.HOLDER```

File: frontend/src/features/settings/components/creditAccount/AddUserModal.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/AddUserModal.tsx]: Either ```console.error(error)``` or navigate to the unexpected error page.

File: frontend/src/features/settings/components/creditAccount/StatusChip.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/StatusChip.tsx:9]: Why not just add "SUSPENDED" as part of the status types enum?

File: frontend/src/features/settings/components/creditAccount/StatusChip.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/StatusChip.tsx]: Extract out the status display string part into a variable, and replace this long className with that variable

File: frontend/src/features/settings/components/creditAccount/UserTable.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/UserTable.tsx]: As mentioned before, use an enum here

File: frontend/src/features/settings/components/creditAccount/UserTable.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/UserTable.tsx]: Can we simplify this boolean expression to just ```return row?.original.userType !== CREDIT_ACCOUNT_USER_TYPES.HOLDER```?

File: frontend/src/features/settings/components/creditAccount/UserTable.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/UserTable.tsx]: I think you need to add handleRemoveButton and userRoles as part of the dependency array as well, and perhaps even better, maybe extract the canUpdateCreditAccount(userRoles) part outside of this callback function, and use its result here instead (and including it in the dependency array)

File: frontend/src/features/settings/components/dashboard/ManageSettingsDashboard.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/dashboard/ManageSettingsDashboard.tsx:25]: Not sure if we still need the feature flags

File: frontend/src/features/settings/hooks/creditAccount.ts
Comment (zgong-gov) [inline frontend/src/features/settings/hooks/creditAccount.ts]: Although it's possible to use getCompanyIdFromSession here inside this hook, I still recommend call it inside the parent component and passing its result into this hook as a param, ie. ```export const useGetCreditAccountQuery = (companyId: string | number) => ...```

File: frontend/src/features/settings/hooks/creditAccount.ts
Comment (zgong-gov) [inline frontend/src/features/settings/hooks/creditAccount.ts]: See my comment on usage of getCompanyIdFromSession from above

File: frontend/src/features/settings/hooks/creditAccount.ts
Comment (zgong-gov) [inline frontend/src/features/settings/hooks/creditAccount.ts]: See my comment on getCompanyIdFromSession from above

File: frontend/src/features/settings/hooks/creditAccount.ts
Comment (zgong-gov) [inline frontend/src/features/settings/hooks/creditAccount.ts]: Similarly here

File: frontend/src/features/settings/hooks/creditAccount.ts
Comment (zgong-gov) [inline frontend/src/features/settings/hooks/creditAccount.ts]: See my comment about getCompanyIdFromSession above

File: frontend/src/features/settings/hooks/creditAccount.ts
Comment (zgong-gov) [inline frontend/src/features/settings/hooks/creditAccount.ts:98]: Performing such tasks in the onSuccess or onError callbacks is possible for now, but I'm not sure about the future if react-query decides to remove these callbacks (like they did for useQuery for the previous version). Just keep this in mind.

File: frontend/src/features/settings/hooks/creditAccount.ts
Comment (zgong-gov) [inline frontend/src/features/settings/hooks/creditAccount.ts]: I feel like you could extract this status logic into a helper method of some sort (preferably put it in the same file as UPDATE_STATUS_ACTIONS), and simply reuse that method call here.

File: frontend/src/features/settings/hooks/creditAccount.ts
Comment (zgong-gov) [inline frontend/src/features/settings/hooks/creditAccount.ts]: Same here

File: frontend/src/features/settings/pages/CreditAccount.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/pages/CreditAccount.tsx]: Use an enum if possible

File: frontend/src/features/settings/pages/CreditAccount.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/pages/CreditAccount.tsx]: Use an enum if possible

File: frontend/src/features/settings/types/CreditAccountActivityColumns.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/types/CreditAccountActivityColumns.tsx]: Why do you need another formatDate method? Does the current ones under "src/common/helpers/formatDate.ts" not accommodate the use cases/needs?

File: frontend/src/features/settings/types/CreditAccountUserColumns.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/types/CreditAccountUserColumns.tsx]: Consider using an enum here

File: frontend/src/features/settings/types/creditAccount.ts
Comment (zgong-gov) [inline frontend/src/features/settings/types/creditAccount.ts]: For these types, use the enum format instead, ie.
```
export const CREDIT_ACCOUNT_TYPES = { SECURED: "SECURED", ... } as const;
export type CreditAccountType = typeof CREDIT_ACCOUNT_TYPES[keyof typeof CREDIT_ACCOUNT_TYPES];
```
and similarly for CreditAccountUserType, CreditAccountStatusType, and CreditAccountActivityType as well

File: database/mssql/scripts/versions/v_28_ddl.sql
Comment (praju-aot) [inline database/mssql/scripts/versions/v_28_ddl.sql:595]: FYI
v_28 is already in prod. Since the change is limited to removing blank spaces, this should not make any impact and is safe to be in v28

File: frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx
Comment (glen-aot) [reply frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx:80]: @krishnan-aot could you clarify?

File: frontend/src/features/settings/components/creditAccount/AccountDetails.tsx
Comment (glen-aot) [reply frontend/src/features/settings/components/creditAccount/AccountDetails.tsx:193]: @zgong-gov yes we are still awaiting completion of the API route, currently blocked by the finance team I believe

Comment (zgong-gov) [reply frontend/src/features/settings/components/creditAccount/AccountDetails.tsx:193]: So I'm wondering if it's even a good idea to create this PR if the API isn't ready... Is that the reason why you included feature flags for credit accounts?

File: frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx
Comment (krishnan-aot) [reply frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx:80]: Credit account isn't complete. Also, we need this feature hidden till we get a green light from relevant teams to expose this feature in production. It is impractical to hold this PR for however long that takes and hence why we can safely merge behind a feature flag without affecting app's state.

Comment (zgong-gov) [reply frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx:80]: @krishnan-aot Ok I see, thanks

File: frontend/src/features/settings/components/creditAccount/AccountDetails.tsx
Comment (zgong-gov) [reply frontend/src/features/settings/components/creditAccount/AccountDetails.tsx:193]: We can discuss whether or not to use these mocked values for credit accounts during the Monday meeting.

File: frontend/src/features/settings/components/dashboard/ManageSettingsDashboard.tsx
Comment (zgong-gov) [reply frontend/src/features/settings/components/dashboard/ManageSettingsDashboard.tsx:25]: Ok, turns out we need feature flags

File: frontend/src/features/settings/components/creditAccount/StatusChip.tsx
Comment (glen-aot) [reply frontend/src/features/settings/components/creditAccount/StatusChip.tsx:9]: @zgong-gov This is because "SUSPENDED" is not a valid Credit Account status type returned from the API, however the styling for the chip that shows the user if a company is suspended shares the same styling as the StatusChip component that displays the credit account status to the user. Rather than create two separate components based on this logic (e.g. CreditAccountStatusChip and SuspendedStatusChip), I prefer to have one shared component like this.

File: frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx
Comment (zgong-gov) [inline frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx]: Extract this into a variable, and use the variable here (ie. ```const companyIdParam = getDefaultRequiredVal(0, companyId)``` and ```useGetCreditAccountQuery(companyIdParam)```)

File: frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx
Comment (zgong-gov) [inline frontend/src/features/manageProfile/components/dashboard/ManageProfilesDashboard.tsx]: Reuse the ```companyIdParam``` variable here

File: frontend/src/features/settings/components/creditAccount/StatusChip.tsx
Comment (zgong-gov) [reply frontend/src/features/settings/components/creditAccount/StatusChip.tsx:9]: ok

File: frontend/src/features/settings/components/creditAccount/UserTable.tsx
Comment (zgong-gov) [reply frontend/src/features/settings/components/creditAccount/UserTable.tsx]: Can you simplify the boolean expression?

File: frontend/src/features/settings/components/creditAccount/AddUserModal.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/AddUserModal.tsx:76]: Do something to handle the error condition, either console.error the error message, or navigate to unexpected error page. Don't leave it and do nothing for the error.

File: frontend/src/features/settings/components/creditAccount/AccountDetails.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/AccountDetails.tsx]: There's no point in printing the status with console.error. Try printing a relevant error message like "Updating credit account status failed or something". This is mostly for dev purposes, showing what action has failed.

File: frontend/src/features/settings/components/creditAccount/AddUserModal.tsx
Comment (zgong-gov) [reply frontend/src/features/settings/components/creditAccount/AddUserModal.tsx:76]: Similarly here, don't print the status, but instead print something more meaningful (eg. "Add user failed" or something).

File: frontend/src/features/settings/components/creditAccount/RemoveUsersModal.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/components/creditAccount/RemoveUsersModal.tsx]: See above

File: frontend/src/features/settings/pages/CreditAccount.tsx
Comment (zgong-gov) [inline frontend/src/features/settings/pages/CreditAccount.tsx]: See above

**Review by zgong-gov — APPROVED** [review]

**Comment by zgong-gov** [comment]
Approved for FE code

**Pull request merged**
